Python/Product/Debugger/DebugAdapterUtils.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using Microsoft.Win32;
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.PythonTools {
+    public class DebugAdapterUtils {
+        public static void GetProcessInfoFromUri(Uri uri, out int pid, out string processName) {
+            ProcessStartInfo psi = new ProcessStartInfo {
+                FileName = @"C:\Windows\System32\NETSTAT.EXE",
+                Arguments = "-a -n -o -p TCP",
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
+                RedirectStandardInput = true,
+                RedirectStandardError = true
+            };
+
+            using (var process = Process.Start(psi)) {
+                process.WaitForExit(250);
+                var portStr = $":{uri.Port}";
+
+                var delimiters = new string[] { " " };
+                while (!process.StandardOutput.EndOfStream) {
+                    var line = process.StandardOutput.ReadLine().ToUpper();
+                    var parts = line.Split(delimiters, StringSplitOptions.RemoveEmptyEntries);
+                    if (parts.Length > 3
+                        && parts[0] == "TCP"
+                        && parts[1].EndsWith(portStr)
+                        && parts[parts.Length - 2] == "LISTENING") {
+                        if (int.TryParse(parts[parts.Length - 1], out pid)) {
+                            var proc = Process.GetProcessById(pid);
+                            processName = proc.ProcessName;
+                            return;
+                        }
+                    }
+                }
+
+                pid = 55555;
+                processName = "python";
+            }
+        }
+
+        public static bool UseExperimentalDebugger() {
+            bool defaultValue = true;
+            if (!_useExperimental.HasValue) {
+                try {
+                    var experimentalKey = @"Software\Microsoft\PythonTools\Experimental";
```

Comment:
This should probably be going through `PythonToolsService.ExperimentalOptions`.

Reply:
Oh, right, it would actually be circular then. Also, the adapter might not even be in VS process (in msvsmon.exe scenarios, for example).

Reply:
You can add a reference to `VSInterpreters`, which has the actual `ExperimentalOptions` class. PythonTools only has it so we can expose them to users - anything checking the options should be going through the lower-level DLL.

Reply:
(And add an `InternalsVisibleTo` to `VSInterpreters` if you need it)